Fix #368 Class cast exception in PluginLoader.loadPlugin

thekingnothing opened the pull request:
Fix issue #368 

- Use `Class.forName` to get plugin implementer class instead loading via context class loader 
- Use `RuntimeException` instead `MockitoException` because `MockitoException` depends on `PluginLoader` and if something incorrect with loader then stack trace  will be also incorrect. 

I don't understand why context class loader was used, because anyway other classes that use plugin is loaded by class loaded of current class.

bric3 commented on src/main/java/org/mockito/internal/configuration/plugins/PluginRegistry.java:
```diff
@@ -8,16 +8,19 @@
     private final MockMaker mockMaker
-            = new PluginLoader(pluginSwitch).loadPlugin(MockMaker.class, "org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker");
+            = new PluginLoader(pluginSwitch).loadPlugin(MockMaker.class, "org.mockito.internal.creation.bytebuddy" +
+                                                                                  ".ByteBuddyMockMaker");
```
Is it possible to avoid the split string ?

bric3 commented on src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java:
```diff
@@ -51,20 +59,22 @@ public PluginLoader(PluginSwitch pluginSwitch) {
         } catch (Exception e) {
             throw new MockitoConfigurationException(
-                    "Failed to load " + service + " implementation declared in " + resources, e);
+                                                           "Failed to load " + service + " implementation declared in " + resources, e);
```
Unnecessary code reformat is Eclipse(/other IDE) acting a fool.

thekingnothing replied:
IDE is configured for another project formatting, in contribute guide I found only mention about spaces. So I've continued use my configuration

bric3 replied:
Yep that's usually an issue ;)
I've myself too in such issue ;) Now I setup my IDE never reformat non modified lines unless I ask it for.

thekingnothing replied:
I've also  setup my IDE to never reformat non modified code, but I face issue then I change code, revert changes, then IDE format code.

bric3 commented on src/test/java/org/mockito/internal/configuration/plugins/PluginLoaderTest.java:
```diff
@@ -0,0 +1,88 @@
+    @Test
+    public void if_custom_context_class_loader_is_used_then_plugin_should_be_load() throws Exception {
+
+        ClassLoader classLoader = ClassLoaders.inMemoryClassLoader().withParent(getParentClassLoader()).withClassDefinition("", readResource()).build();
```
This doesn't work as expected, it is like creating a noname class, whose class byte code is the string found in the mockito-extensions file ;)

thekingnothing replied:
If you look at source code of  `InMemoryClassLoader`

 ```
@Override
        public Enumeration<URL> getResources(String ignored) throws IOException {
            return inMemoryOnly();
        }

        private Enumeration<URL> inMemoryOnly() {
            final Set<String> names = inMemoryClassObjects.keySet();
            return new Enumeration<URL>() {
                private final MemHandler memHandler = new MemHandler(InMemoryClassLoader.this);
                private final Iterator<String> it = names.iterator();

                public boolean hasMoreElements() {
                    return it.hasNext();
                }

                public URL nextElement() {
                    try {
                        return new URL(null, SCHEME + ":" + it.next(), memHandler);
                    } catch (MalformedURLException rethrown) {
                        throw new IllegalStateException(rethrown);
                    }
                }
            };
        }
```

The `InMemoryClassLoader` returns all class as resources and `PluginLoader` reads these resources. So if I remove `readResource` then `PluginLoader` will not find any plugin and then create default instance. 

I can replace `readResource` via `"org.mockito.internal.configuration.plugins.PluginImpl".getBytes()`, but I prefer use same arrangement as for another test.

bric3 commented on src/test/java/org/mockito/internal/configuration/plugins/PluginLoaderTest.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Test
+    public void without_custom_context_class_loader_plugin_should_be_load() {
```
Could be renamed to `should_load_custom_plugin_without_context_classloader`

bric3 commented on src/test/java/org/mockito/internal/configuration/plugins/PluginLoaderTest.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Test
+    public void if_custom_context_class_loader_is_used_then_plugin_should_be_load() throws Exception {
```
Could be renamed to `should_load_custom_plugin_with_custom_context_classloader`

bric3 commented on src/test/java/org/mockito/internal/configuration/plugins/PluginLoaderTest.java:
```diff
@@ -0,0 +1,88 @@
+        Object pluginInterface = pluginLoader.loadPlugin(PluginInterface.class, "org.mockito.internal" +
+                                                                                                  ".configuration.plugins.DefaultPlugin");
+
+        assertEquals(PluginImpl.class, pluginInterface.getClass());
```
Asserts could be more useful when asserting on classloader themself in this case :
```java
assertEquals(classLoader.loadClass(PluginImpl.class.getName()).getClassLoader(), pluginInterface.getClass().getClassLoader());
```

thekingnothing replied:
Suggested assert will be incorrect.  The `PluginImpl`should be loaded via `classLoader` (who plays role context class loader). It should be loaded with `AppClassLoader` to be assignable to fields of `PluginRegistry` even if context class loader != `AppClassLoader`. I change assert to more readable.

thekingnothing commented:
@bric3, 

I modified test. I renamed test methods, change comment so it now more correspond to my goal and what I really test. I change assert and give meaningful name. Also I fixed another mistake - I used Mockito Annotation to test PluginLoader then annotation injector uses PluginLoader. I don't use Mockito any more to test Mockito :)

bric3 commented on src/test/java/org/mockito/internal/configuration/plugins/PluginLoaderTest.java:
```diff
@@ -0,0 +1,87 @@
+    public void should_load_custom_plugin_with_custom_context_classloader() throws Exception {
+
+        ClassLoader classLoader = ClassLoaders.inMemoryClassLoader().withParent(getParentClassLoader())
+                                          .withClassDefinition("", readResource()).build();
```
This is still incorrect use of the in-memory classloader ;)
`.withClassDefinition("", readResource())` has no impact on test

thekingnothing replied:
If I remove `.withClassDefinition("", readResource())` then test will fail, because default plugin implementer will be loaded. Without class definition the in-memory classloader returns empty resource `Enumeration` . So I need a easy way to create class loader which will return as resource that I need to return and failed to load any class. Any suggestion?

bric3 replied:
This is definitely wrong. The `.withClassDefinition("", readResource())` directive registers a the class name `""` whose class byte definition is the content of the file `mockito-extensions/org.mockito.internal.configuration.plugins.PluginInterface`. If this code was used then the JVM would raise an error.

thekingnothing replied:
The `.withClassDefinition("", readResource())` doesn't register the class. It simple put value of `
readResource()` to map with key empty string as key. 
`private Map<String , byte[]> inMemoryClassObjects = new HashMap<String , byte[]>();`

And it's used only in to cases: 

-  find class with name, but empty string is invalid class name value, so it should never be called and if it called then `Exception`is correct behaviour for this case 
- getResources - then `Enumeration` is return. 

I could use not empty string, but `"mockito-extensions/org.mockito.internal.configuration.plugins.PluginInterface"`, but meaning will be same.

bric3 replied:
I understand the idea, but the test pass with or without `.withClassDefinition("", readResource())`.

The `ClassLoaders` is just a basic tool that wasn't designed to support something else but class bytes. It may be something to improve. Anyway it's just coincidental that file is supported.

Anyway for better readability it would have been clearer to write it like that, and then it's possible to discard `readResource()` :

```java
ClassLoader classLoader = ClassLoaders.inMemoryClassLoader()
                                      .withParent(getParentClassLoader())
                                      .withClassDefinition("mockito-extensions/org.mockito.internal.configuration.plugins.PluginInterface", 
                                                           PluginImpl.class.getName().getBytes("UTF-8"))
                                      .build();
```

thekingnothing replied:
Test passed without because assertion was a bit incorrect after last refactoring :(
I fixed assertion and added one more test which check loading default plugin.

bric3 commented on src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java:
```diff
@@ -51,20 +59,22 @@ public PluginLoader(PluginSwitch pluginSwitch) {
             String foundPluginClass = new PluginFinder(pluginSwitch).findPluginClass(Iterables.toIterable(resources));
             if (foundPluginClass != null) {
-                Class<?> pluginClass = loader.loadClass(foundPluginClass);
+                Class<?> pluginClass = Class.forName(foundPluginClass);
```
What do you think of trying first `loader.loadClass(found)` then if it doesn't work try the `Class.forName(found)` ?
I'm wary of the plain `Class.forName` as it will always choose the mockito current classloader.

bric3 replied:
After playing a bit with the code, all the tests can still pass with the old statement `loader.loadClass(found)`

thekingnothing replied:
I wonder how  old statement could work in this case: 

1. The `MockUtil` is created first time from class which is loaded by Application class loader.  Important that no other Mockito classes haven been used before creating `MockUtil`. So it means that the `MockUtil` class will be loaded by this class loader. 
2. During class initialisation the `Plugins.getMockMaker` is  called, an `Plugins` class also loaded by application class loader.
3. Next step `PluginRegistry` - and again application class loader is used. 
4. And now the `PluginLoader` is called, which uses completed different context class loader to load plugin. Okay, plugin can be loaded and may even cast and return, but how it could be assign to field? 

By application class loader I mean that it's class loader which is used to load first class with `main()`method. A  context class loader - is a new class loader which is create via `new` and doesn't have application class loader as a parent

raphw replied:
Depends on the context class loader which could be just any class loader. (Even though tests are mainly run from the class path.) I do however agree that this should be `loader.findClass` at this place.

bric3 commented on src/test/resources/mockito-extensions/org.mockito.internal.configuration.plugins.PluginInterface:
```diff
@@ -0,0 +1 @@
+org.mockito.internal.configuration.plugins.PluginImpl
```
For documentation maybe add 

```
# Used for test : org.mockito.internal.configuration.plugins.PluginLoaderTest
```

thekingnothing replied:
Done

bric3 commented on src/test/java/org/mockito/internal/configuration/plugins/PluginLoaderTest.java:
```diff
@@ -0,0 +1,76 @@
+        boolean assertion = DefaultPlugin.class.isAssignableFrom(plugin.getClass());
+        String message = String.format("Plugin class %s is not assignable to default %s", plugin.getClass(), DefaultPlugin.class);
+        assertTrue(message, assertion);
+    }
```
If you want there's AssertJ on the classpath, that could be rewritten to 

```java
Assertions.assertThat(plugin.getClass())
          .describedAs("Loaded plugin class '%s' is not assignable to local class '%s'", plugin.getClass(), PluginImpl.class)
          .isAssignableFrom(PluginImpl.class);
```

raphw commented:
I do not think that this PR is correct. The basic change is to disable that a plugin is loaded via a class loader that determines the plugin but via Mockito's class loader. If Mockito is loaded via an OSGi bundle, for example, this would break visibility constraints. 

We need to require that a class loader that identifies a plugin can also load this plugin and can also see Mockito. For the tests, this is broken.  

It is however true that the exceptions should not rely on the plugin architecture. I will fix this part manually. Thank you for opening this PR, please get in touch if you need to discuss this issue further.
